tracer/test/Datadog.Trace.Security.Unit.Tests/IAST/Tainted/DefaultTaintedMapTests.cs

```diff
@@ -0,0 +1,337 @@
+    {
+        List<string> objects = new();
+        DefaultTaintedMap map = new();
+        AssertFlatMode(map, objects);
+
+        foreach (var itemInMap in map.GetListValues())
+        {
+            Assert.Contains((itemInMap as TaintedObject).Value, objects);
+        }
+    }
+
+    [Fact]
+    public void GivenATaintedObjectMap_WhenPutMoreThanDefaultFlatModeThresoldElements_GetsFlatAndOnlyOneObjectWithSameHashIsStored()
+    {
+        List<string> objects = new();
+        DefaultTaintedMap map = new();
+        AssertFlatMode(map, objects);
+
+        var testString = new StringForTest(Guid.NewGuid().ToString());
+        testString.Hash = 10;
+        map.Put(new TaintedForTest(testString, null));
+
+        Assert.NotNull(map.Get(testString));
+
+        var testString2 = new StringForTest(Guid.NewGuid().ToString());
+        testString2.Hash = 10;
+        map.Put(new TaintedForTest(testString2, null));
+
+        Assert.NotNull(map.Get(testString2));
+        Assert.Null(map.Get(testString));
+    }
+
+    [Fact]
+    public void GivenATaintedObjectMap_WhenFlatMode_LastElementIsStored()
+    {
+        List<string> objects = new();
+        DefaultTaintedMap map = new();
+        AssertFlatMode(map, objects);
+
+        for (int i = 0; i < 100; i++)
+        {
+            var testString = new StringForTest(Guid.NewGuid().ToString());
+            map.Put(new TaintedForTest(testString, null));
+            Assert.NotNull(map.Get(testString));
+        }
+    }
+
+    [Fact]
+    public void GivenATaintedObjectMap_WhenPutObjectsThatGetDisposed_ObjectsArePurged()
+    {
+        DefaultTaintedMap map = new();
+        List<string> disposedObjects = new();
+        List<string> aliveObjects = new();
+        bool alive = true;
+
+        for (int i = 0; i < DefaultTaintedMap.DefaultFlatModeThresold / 2; i++)
+        {
+            var testString = Guid.NewGuid().ToString();
+            var tainted = new TaintedForTest(testString, null);
+            map.Put(tainted);
+            (alive ? aliveObjects : disposedObjects).Add(testString);
+            alive = !alive;
+        }
+
+        AssertContained(map, aliveObjects);
+        AssertContained(map, disposedObjects);
+
+        foreach (var item in disposedObjects)
+        {
+            (map.Get(item) as TaintedForTest).SetAlive(false);
+        }
+
+        map.Purge();
+        AssertContained(map, aliveObjects);
+        AssertNotContained(map, disposedObjects);
+    }
+
+    [Fact]
+    public void GivenATaintedObjectMap_WhenASingleObjectNotDisposed_IsNotPurged()
+    {
+        DefaultTaintedMap map = new();
+        var testString = Guid.NewGuid().ToString();
+        var tainted = new TaintedForTest(testString, null);
+        map.Put(tainted);
+        Assert.NotNull(map.Get(testString));
+        map.Purge();
+        Assert.NotNull(map.Get(testString));
+    }
+
+    [Fact]
+    public void GivenATaintedObjectMap_WhenASingleObjectDisposed_IsPurged()
```

Comment:
I think there's one test case not covered:

When you `Put` more objects than the `PurgeCount`, that it automatically triggers a purge. Could add `PurgeCount - 1` objects, and confirm the inner list contains them all, then add one more and confirm the list is purged?

Reply:
PurgeCount controls the probability of an object insertion to cause a purge based on it's hash id. A test can done, based on you suggestion, by controlling the hash ids of the objects, so objects that will not cause a purge can be added and, then, add one that will cause it, so the automatic purge mechanism can be covered. I will test that.